### DISCUSSION TITLE
Please don't deprecate subscribe

### DISCUSSION THREAD

**Discussion by omasback** [pmndrs/zustand#1433] [General]
`subscribe` is the reason I chose zustand a couple years ago. So I could subscribe some react-spring animated components to high frequency store updates without triggering react rerenders. I thought it was a critical part of the library because 0xca0a/@ drcmda used it for high performance graphics applications. Why was it removed from the core? Now I'm trying to update the lib from 3.5.10 and the amount of exotic type annotations I've been trial-and-erroring to get subscribeWithSelector to work is maddening. And I still can't get it to work! 

And the type of store that's returned from  vanilla `createStore` is `Write<StoreApi<StoreState>, StoreSubscribeWithSelector<StoreState>>`, but I can't actually use `Write` and  `StoreSubscribeWithSelector` types elsewhere in my code because they arent exported from the library. 

And what is the point anyways? So I have to add a middleware to every store and invert the order of arguments to the subscribe method (why?). And for what? Just to keep current? Seems like this library is just getting harder to use, without any big improvements to justify the work of updating. Not sure if I'm asking for anything in particular here. I will probably just stay on version 3 forever. Maybe it would be nice to be able to choose a version on the documentation site for those of us who choose not to upgrade.

**Comment by omasback** [comment]
And to be clear I like the new "vanilla" store option. I'm currently passing the original store hook down via context and I have to be careful in my code to not change the reference to ensure hook stability. @dai-shi and I tried to fix that issue a while back with the custom `createContext` function but I guess that's fallen by the wayside, which is fine. I understand why it was insufficient. Maybe what I need to do is create a "zustand-classic" type of package based on a fork of V3. Which maybe could include a vanilla store option which would be just the api that could be consumed by a useStore hook. But it wouldn't require a lot of refactoring like v4. idk tbh I really don't have the time to do that. But a man can dream, can't he?

**Reply by dai-shi** [reply]
btw, creating a zustand-like library is pretty easy.
- I made a course about it: https://egghead.io/courses/create-a-zustand-like-module-state-library-bf55241e
- And this: https://daishi.gumroad.com/l/uaxms

People can easily create a zustand clone or a similar functionality, especially in JS.
That's actually the reason why we focus on providing full TS support in v4. That would be a unique value.

**Comment by dai-shi** [comment]
Please upgrade to the latest version of v3, before updating to v4.
I assume your concern is TS specific, because there's basically no difference in JS.

TS in v4 is very tricky, but it gives you better type support. It's a trade off.
If you can provide a dummy small reproduction, we should be able to help.

**Comment by dai-shi** [comment]
About `subscribeWithSelector` middleware, it's specifically provided for people like you migrating existing code base. So, it's unfortunate to hear your bad experience. (Note that the API is not compatible, because the order of parameters is changed.)

However, for people who use zustand from now, I don't recommend it.
In general, middleware is hard to type with, and it's trivial to give the same functionality without it.
Recently, I found a good code snippet to demonstrate it:
```js
const useFooStore = create(...)
useFooStore.subscribeWithSelector = (selector, callback) => useFooStore.subscribe((currState, prevState) => {
  const currSelected = selector(currState)
  const prevSelected = selector(prevState)
  if (!Object.is(currSelected, prevSelected)) {
    callback(currSelected, prevSelected)
  }
}
```
